server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregatorTests.java

```diff
@@ -0,0 +1,390 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.histogram;
+
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.SortedNumericDocValuesField;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.RandomIndexWriter;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.NumericUtils;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.NumberFieldMapper;
+import org.elasticsearch.search.aggregations.AggregationBuilders;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorTestCase;
+import org.elasticsearch.search.aggregations.bucket.terms.Terms;
+import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregationBuilder;
+import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregator;
+import org.elasticsearch.search.aggregations.metrics.InternalStats;
+import org.elasticsearch.search.aggregations.metrics.StatsAggregationBuilder;
+import org.elasticsearch.search.aggregations.support.AggregationInspectionHelper;
+import org.elasticsearch.search.aggregations.support.ValueType;
+
+import java.io.IOException;
+
+public class VariableWidthHistogramAggregatorTests extends AggregatorTestCase {
+
+    public void testNoDocs() throws Exception{
+        try (Directory dir = newDirectory();
```

Comment:
We've been moving most of this stuff to the `testCase` method from `AggregatorTestCase`. Or the `search` or `searchAndReduce` methods. They do some randomization which is nice. `testCase` and `searchAndReduce` do random partial reduces and other interesting things that are helpful in validating that your aggregator makes sense combined with the reduce implementation on `InternalVariableWidthAggregation`. `search` is pretty much what you are doing here so you can still assert exact stuff about the shape of things that your aggregator produces.

I've seen the tests for `auto_date_histogram` define test data and request one time and then run a `search` case and assert about the shape of the results directly from the aggregator and then run a `searchAndReduce` case and assert that the reduction is perfect.

Yes, this kind of overlaps with the stuff in `InternalAggregationTestCase`. I do think it is nice to test both things together on real documents and real lucene indices like you get out of this class. I'm not super comfortable with the overlap though. :shrug:

Reply:
So far I've migrated all the tests, except for the ones with sub aggregations, to use `search` and `searchAndReduce`. But I'm seeing some results that are a little strange, due to the randomization.

Problem #1: The aggregation caches the first `cache_limit` docs and initializes the first set of buckets with them. So, it the document order is randomized then the final buckets might sometimes be different.

Problem #2: Unlike most histogram aggregations, if we have 3 variable width histograms A, B, C: `reduce([reduce([A, B]), C]) != reduce([A, reduce([B,C])])`. This is because the closest two buckets are iteratively merged. A different set of histograms being reduced first means that a different set of buckets will be merged together early. This will cause different centroids and change the buckets that are merged later. In contrast, the other histograms have fixed intervals so the reduce order is irrelevant.

Because of these two problems, the randomization very occasionally makes the tests fail. I currently call both `search` and `searchAndReduce` throughout the tests, and if I run all the tests with `-Dtests.iters=100`, they will fail ~2% of the time. If I remove all the calls to `searchAndReduce` (which removes the problem #2), the tests will fail ~0.7% of the time.

I can live with this, as I know that it's the randomization that's causing it. Do you think this is a problem?

Reply:
I don't think we could commit with known randomized failures, but I understand the problem.

Number 1 might be worth working around by forcing a single segment - you don't get all the lovely randomization, but you still get some. You can force a single segment by using `IndexWriter#addDocuments` to add all the documents at once.

It might be worth adding a test that *doesn't* force a single segment with much looser assertions about the result. That is sort of the idea behind `assertReduced`, but running in a different place.

For problem 2 that is interesting! We've intentionally stayed away from aggs that don't have that commutative property of reduction in the past. And, yeah, you sort of can't. Not without forcing it by doing funny stuff like sorting the resulting aggregations by something fixed. And skipping partial reductions too..... I'm gonna need to think about this some more!

Reply:
> Hi Nik, sorry for the delay.

I feel like I've been so slow responding to you!

OK! I think we're just going to have to be ok with the non-commutative nature of the beast.

Reply:
Okay sounds good! Any thoughts on next steps for this PR?